src/DataLayer/dataSources/DataSource.ts

```diff
@@ -1,25 +1,95 @@
 // (C) 2007-2018 GoodData Corporation
-import { AFM } from '@gooddata/typings';
+import { AFM, Execution } from '@gooddata/typings';
 import * as stringify from 'json-stable-stringify';
 import { IDataSource } from '../interfaces/DataSource';
 
 export {
     IDataSource
 };
 
+export const SERVER_SIDE_COLUMN_LIMIT = 1000;
+export const SERVER_SIDE_ROW_LIMIT = 1000;
+
+const ROW_DIMENSION_INDEX = 0;
+const COLUMN_DIMENSION_INDEX = 1;
+
 export type ExecFactory<T> = (resultSpec: AFM.IResultSpec) => Promise<T>;
 
 export class DataSource<T> implements IDataSource<T> {
+    private executionPromises: {
+        [key: string]: Promise<Execution.IExecutionResponse>
+    };
+
     constructor(
         private execFactory: ExecFactory<T>,
         private afm: AFM.IAfm,
-        private fingerprint?: string
-    ) {}
+        private fingerprint?: string,
+        private responseFactory?: (resultSpec: AFM.IResultSpec) => Promise<Execution.IExecutionResponse>,
+        private resultFactory?:
+            (executionResultUri: string, offset: number[], limit: number[]) => Promise<Execution.IExecutionResult>
+    ) {
+        this.executionPromises = {};
+    }
 
     public getData(resultSpec: AFM.IResultSpec): Promise<T> {
         return this.execFactory(resultSpec);
     }
 
+    public getPage(
+        resultSpec: AFM.IResultSpec,
+        offset: number[] = [],
+        limit: number[] = []
+    ): Promise<Execution.IExecutionResponses> {
+        const resultSpecFingerprint = stringify(resultSpec);
+        if (!this.responseFactory) {
+            throw new Error('Missing responseFactory!');
```

Comment:
this would not be necessary with `PageableDataSource`

Reply:
Due to discussion with @tomasreichmann it will be quite difficult to change it and keep backward compatibility